File: Commons/Graph/src/main/java/com/greentree/commons/graph/algorithm/walk/ChildrenWalker.java
package com.greentree.commons.graph.algorithm.walk;

import com.greentree.commons.graph.RootTree;

public class ChildrenWalker<V> implements RootTreeWalker<V> {
	
	private final RootTree<? extends V> tree;
	
	public ChildrenWalker(RootTree<? extends V> tree) {
		this.tree = tree;
	}
	
	
	@Override
	public void visit(V start, GraphVisitor<? super V> visitor) {
		walk(visitor, start);
	}
	
	private void walk(GraphVisitor<? super V> visitor, V v) {
		visitor.startVisit(v);
		for(var to : tree.getChildren(v))
			walk(visitor, to, v);
		visitor.endVisit(v);
	}
	
	private void walk(GraphVisitor<? super V> visitor, V v, V parent) {
		visitor.startVisit(parent, v);
		for(var to : tree.getChildren(v))
			walk(visitor, to, v);
		visitor.endVisit(parent, v);
	}
	
	@Override
	public RootTree<? extends V> graph() {
		return tree;
	}
	
}

File: Commons/Graph/src/main/java/com/greentree/commons/graph/algorithm/walk/ChildrenWalker.kt
package com.greentree.commons.graph.algorithm.walk

import com.greentree.commons.graph.RootTree
import com.greentree.commons.graph.algorithm.walk.vertex.VertexVisitor

class ChildrenWalker<V : Any>(private val tree: RootTree<V>) :
	RootTreeWalker<V> {

	override fun visit(start: V, visitor: VertexVisitor<V>) {
		walk(visitor, start)
	}

	override fun graph(): RootTree<V> {
		return tree
	}

	private fun walk(visitor: VertexVisitor<V>, v: V) {
		visitor.startVisit(v)
		for(to in tree.getChildren(v)) walk(visitor, to, v)
		visitor.endVisit(v)
	}

	private fun walk(visitor: VertexVisitor<V>, v: V, parent: V) {
		visitor.startVisit(parent, v)
		for(to in tree.getChildren(v)) walk(visitor, to, v)
		visitor.endVisit(parent, v)
	}
}
